src/web/extractors.rs

```diff
@@ -168,6 +368,81 @@ impl FromRequest<ServerState> for CollectionRequest {
     }
 }
 
+/// Collection Request Post extractor
+///
+/// Extracts/validates information needed for batch collection POST requests.
+pub struct CollectionPostRequest {
+    pub collection: String,
+    pub state: State<ServerState>,
+    pub user_id: HawkIdentifier,
+    pub query: BsoQueryParams,
+    pub bsos: BsoBodies,
+}
+
+impl FromRequest<ServerState> for CollectionPostRequest {
+    type Config = ();
+    type Result = Box<Future<Item = CollectionPostRequest, Error = Error>>;
+
+    /// Extractor for Collection Posts (Batch BSO upload)
+    ///
+    /// Utilizes the `BsoBodies` for parsing, and add's two validation steps not
+    /// done previously:
+    ///   - If the collection is 'crypto', known bad payloads are checked for
+    ///   - Any valid BSO's beyond `BATCH_MAX_RECORDS` are moved to invalid
+    fn from_request(req: &HttpRequest<ServerState>, _: &Self::Config) -> Self::Result {
+        let fut = <(
+            HawkIdentifier,
+            State<ServerState>,
+            CollectionParam,
+            BsoQueryParams,
+            BsoBodies,
+        )>::extract(req).and_then(|(user_id, state, collection, query, mut bsos)| {
+            let collection = collection.collection.clone();
+            if collection == "crypto" {
+                // Verify the client didn't mess up the crypto if we have a payload
+                for bso in bsos.valid.iter() {
+                    if let Some(ref data) = bso.payload {
+                        if KNOWN_BAD_PAYLOAD_REGEX.is_match(data) {
+                            return future::err(
+                                ValidationErrorKind::FromDetails(
+                                    "Known-bad BSO payload".to_owned(),
+                                    RequestErrorLocation::Body,
+                                    Some("bsos".to_owned()),
+                                ).into(),
+                            );
+                        }
+                    }
+                }
+            }
+
+            // Trim the excess BSO's to be under the batch size
+            let overage: i64 = (bsos.valid.len() as i64) - (BATCH_MAX_RECORDS as i64);
+            if overage > 0 {
+                for _ in 1..=overage {
```

Comment:
it probably doesn't matter much but you could use Vec::drain here instead while maintaining the original order of the invalids

```rust
                bsos.invalid
                    .extend(bsos.valid.drain(overage as usize..).map(|last| BsoBody {
                        id: Some(last.id),
                        sortindex: last.sortindex,
                        payload: last.payload,
                        ttl: last.ttl,
                    }));
```